In [1]:
import pandas as pd

# Reading in the dataset

In [5]:
stats=pd.read_csv("player_mvp_stats.csv")

In [6]:
stats.head(50)

Unnamed: 0           Player Pos  Age   Tm   G  GS    MP   FG   FGA  ...  \
0            0       A.C. Green  PF   27  LAL  82  21  26.4  3.1   6.6  ...   
1            1      Byron Scott  SG   29  LAL  82  82  32.1  6.1  12.8  ...   
2            2   Elden Campbell  PF   22  LAL  52   0   7.3  1.1   2.4  ...   
3            3    Irving Thomas  PF   25  LAL  26   0   4.2  0.7   1.9  ...   
4            4     James Worthy  SF   29  LAL  78  74  38.6  9.2  18.7  ...   
5            5       Larry Drew  PG   32  LAL  48   2  10.3  1.1   2.6  ...   
6            6    Magic Johnson  PG   31  LAL  79  79  37.1  5.9  12.4  ...   
7            7  Mychal Thompson   C   36  LAL  72   4  15.0  1.6   3.2  ...   
8            8      Sam Perkins  PF   29  LAL  73  66  34.3  5.0  10.2  ...   
9            9     Terry Teagle  SG   30  LAL  82   0  18.3  4.1   9.2  ...   
10          10       Tony Smith  SG   22  LAL  64   1  10.9  1.5   3.4  ...   
11          11      Vlade Divac   C   22  LAL  82  81  28.2  4.4   7.8  ...   
12          12       A.C. Green  PF   28  LAL  82  53  35.4  4.7   9.8  ...   
13          13      Byron Scott  SG   30  LAL  82  82  32.7  5.6  12.3  ...   
14          14     Chucky Brown  SF   23  LAL  42   2  10.3  1.4   3.0  ...   
15          15   Cliff Robinson  PF   31  LAL   9   0   8.7  1.2   3.0  ...   
16          16  Demetrius Calip  PG   22  LAL   7   0   8.3  0.6   2.6  ...   
17          17   Elden Campbell   C   23  LAL  81  47  23.2  2.7   6.1  ...   
18          18       Jack Haley   C   28  LAL  49   9   8.0  0.6   1.7  ...   
19          19     James Worthy  SF   30  LAL  54  54  39.0  8.3  18.6  ...   
20          20      Keith Owens  SF   22  LAL  20   0   4.0  0.5   1.6  ...   
21          21     Rory Sparrow  PG   33  LAL  46   0  10.6  1.3   3.3  ...   
22          22      Sam Perkins  PF   30  LAL  63  63  37.0  5.7  12.7  ...   
23          23   Sedale Threatt  PG   30  LAL  82  82  37.4  6.2  12.7  ...   
24          24     Terry Teagle  SG   31  LAL  82   0  19.5  4.4   9.8  ...   
25          25       Tony Smith  PG   23  LAL  63   0  13.0  1.8   4.5  ...   
26          26      Vlade Divac   C   23  LAL  36  18  27.2  4.4   8.8  ...   
27          27       A.C. Green  PF   29  LAL  82  55  34.4  4.6   8.6  ...   
28          28   Alex Blackwell  SF   22  LAL  27   0   4.0  0.5   1.6  ...   
29          29   Anthony Peeler  SG   23  LAL  77  11  21.5  3.9   8.2  ...   
30          30  Benoit Benjamin   C   28  LAL  59   6  12.8  2.3   4.6  ...   
31          31      Byron Scott  SG   31  LAL  58  53  28.9  5.1  11.4  ...   
32          32    Doug Christie  SG   22  LAL  23   0  14.4  2.0   4.6  ...   
33          33     Duane Cooper  PG   23  LAL  65   0   9.9  1.0   2.4  ...   
34          34   Elden Campbell  PF   24  LAL  79  13  19.6  3.0   6.6  ...   
35          35    James Edwards   C   37  LAL  52   0  11.9  2.3   5.2  ...   
36          36     James Worthy  SF   31  LAL  82  69  28.8  6.2  13.9  ...   
37          37   Sedale Threatt  PG   31  LAL  82  82  35.3  6.4  12.5  ...   
38          38       Tony Smith  SG   24  LAL  55   9  13.7  2.4   5.0  ...   
39          39      Vlade Divac   C   24  LAL  82  69  30.8  4.8  10.0  ...   
40          40       A.C. Green  PF   30  PHO  82  55  34.5  5.7  11.3  ...   
41          41  Cedric Ceballos  SF   24  PHO  53  43  30.2  8.0  15.0  ...   
42          42  Charles Barkley  PF   30  PHO  65  65  35.4  8.0  16.1  ...   
43          43      Dan Majerle  SG   28  PHO  80  76  40.1  6.0  14.2  ...   
44          44      Danny Ainge  SG   34  PHO  68   1  22.9  3.3   7.9  ...   
45          45     Duane Cooper  PG   24  PHO  23   2   5.9  0.8   1.8  ...   
46          46     Elliot Perry  PG   24  PHO  27   9  16.0  1.6   4.2  ...   
47          47    Frank Johnson  PG   35  PHO  70   5  12.5  1.9   4.3  ...   
48          48    Jerrod Mustaf  PF   24  PHO  33   2   5.9  0.9   2.5  ...   
49          49       Joe Kleine   C   32  PHO  74   4 

In [7]:
del stats["Unnamed: 0"]

In [8]:
#Removing the Null Values

stats=stats.fillna(0)

# Getting the parameters which will be useful for MVP prediction

In [10]:
stats.columns

Index(['Player', 'Pos', 'Age', 'Tm', 'G', 'GS', 'MP', 'FG', 'FGA', 'FG%', '3P',
       '3PA', '3P%', '2P', '2PA', '2P%', 'eFG%', 'FT', 'FTA', 'FT%', 'ORB',
       'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS', 'Year',
       'Pts Won', 'Pts Max', 'Share', 'Team', 'W', 'L', 'W/L%', 'GB', 'PS/G',
       'PA/G', 'SRS'],
      dtype='object')

In [11]:
predictors=['Age', 'G','GS','MP', 'FG', 'FGA', 'FG%', '3P',
       '3PA', '3P%', '2P', '2PA', '2P%', 'eFG%', 'FT', 'FTA', 'FT%', 'ORB',
       'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS', 'Year','W', 'L', 'W/L%', 'GB', 'PS/G',
       'PA/G', 'SRS']

# Segregating the dataset into training and testing data

In [12]:
train = stats[stats["Year"]<2020]

In [13]:
test = stats[stats["Year"]==2020]

# Using Ridge regression 

In [15]:
from sklearn.linear_model import Ridge 

reg = Ridge (alpha=.1)

In [16]:
reg.fit(train[predictors],train["Share"])

Ridge(alpha=0.1)

# Predicting the Share of MVP 

In [17]:
predictions = reg.predict(test[predictors])

In [19]:
predictions= pd.DataFrame(predictions,columns=["predictions"],index=test.index)

In [20]:
predictions 

predictions
611       0.020689
612      -0.028052
613       0.000923
614      -0.014191
615       0.008199
...            ...
12554     0.004809
12555     0.037983
12556     0.007636
12557     0.009623
12558     0.009548

[529 rows x 1 columns]

In [21]:
combination = pd.concat([test[["Player","Share"]],predictions],axis=1)

In [25]:
combination

Player  Share  predictions
611               Aaron Gordon    0.0     0.020689
612            Al-Farouq Aminu    0.0    -0.028052
613            Amile Jefferson    0.0     0.000923
614               B.J. Johnson    0.0    -0.014191
615              D.J. Augustin    0.0     0.008199
...                        ...    ...          ...
12554  Rondae Hollis-Jefferson    0.0     0.004809
12555              Serge Ibaka    0.0     0.037983
12556           Shamorie Ponds    0.0     0.007636
12557          Stanley Johnson    0.0     0.009623
12558            Terence Davis    0.0     0.009548

[529 rows x 3 columns]

In [27]:
combination.sort_values("Share",ascending=False).head(10)

Player  Share  predictions
11848  Giannis Antetokounmpo  0.952     0.239042
3581            LeBron James  0.746     0.161240
9307            James Harden  0.363     0.187201
8954             Luka Dončić  0.198     0.168907
5116           Kawhi Leonard  0.166     0.131559
3569           Anthony Davis  0.081     0.144882
1316              Chris Paul  0.026     0.079222
7195          Damian Lillard  0.023     0.132910
10938           Nikola Jokić  0.018     0.102292
12551          Pascal Siakam  0.017     0.077603

# Finding an error metric for the MVP problem

In [28]:
from sklearn.metrics import mean_squared_error

In [29]:
mean_squared_error(combination["Share"], combination["predictions"])

0.002524802124295883

In [30]:
combination["Share"].value_counts()

0.000    517
0.026      1
0.081      1
0.746      1
0.166      1
0.009      1
0.023      1
0.198      1
0.363      1
0.018      1
0.001      1
0.952      1
0.017      1
Name: Share, dtype: int64

In [31]:
#Assigning a Rank
combination=combination.sort_values("Share",ascending=False)
combination["Rk"]=list(range(1,combination.shape[0]+1))

In [32]:
combination.head(10)

Player  Share  predictions  Rk
11848  Giannis Antetokounmpo  0.952     0.239042   1
3581            LeBron James  0.746     0.161240   2
9307            James Harden  0.363     0.187201   3
8954             Luka Dončić  0.198     0.168907   4
5116           Kawhi Leonard  0.166     0.131559   5
3569           Anthony Davis  0.081     0.144882   6
1316              Chris Paul  0.026     0.079222   7
7195          Damian Lillard  0.023     0.132910   8
10938           Nikola Jokić  0.018     0.102292   9
12551          Pascal Siakam  0.017     0.077603  10

In [46]:
combination =combination.sort_values("predictions",ascending=False)
combination["Predicted_Rk"]=list(range(1,combination.shape[0]+1))

In [47]:
combination.head(10)

Player  Share  predictions   Rk  Predicted_Rk
11848  Giannis Antetokounmpo  0.952     0.239042    1             1
9307            James Harden  0.363     0.187201    3             2
8954             Luka Dončić  0.198     0.168907    4             3
3581            LeBron James  0.746     0.161240    2             4
3569           Anthony Davis  0.081     0.144882    6             5
7195          Damian Lillard  0.023     0.132910    8             6
5116           Kawhi Leonard  0.166     0.131559    5             7
2480             Joel Embiid  0.000     0.124611  300             8
4226      Karl-Anthony Towns  0.000     0.123287  508             9
11268             Trae Young  0.000     0.118689  265            10

In [54]:
def find_ap(combination):
    actual=combination.sort_values("Share",ascending=False).head(5)
    predicted=combination.sort_values("predictions",ascending=False)
    ps=[]
    found=0
    seen=1
    for index, row in predicted.iterrows():
        if row["Player"] in actual["Player"].values:
            found+=1
            ps.append(found/seen)
        seen+=1
    return sum(ps)/len(ps) ## Error Metric

## Error Metric:

In [55]:
find_ap(combination)

0.9428571428571428

In [56]:
years = list(range(1991,2021))

In [62]:
aps = []
all_predictions = []
for year in years[5:]:
    train = stats[stats["Year"] < year]
    test = stats[stats["Year"] == year]
    reg.fit(train[predictors],train["Share"])
    predictions = reg.predict(test[predictors])
    predictions = pd.DataFrame(predictions, columns=["predictions"], index=test.index)
    combination = pd.concat([test[["Player", "Share"]], predictions], axis=1)
    all_predictions.append(combination)
    aps.append(find_ap(combination))

In [63]:
sum(aps) / len(aps)

0.7091945189766616

In [57]:
def add_ranks(predictions):
    predictions = predictions.sort_values("predictions", ascending=False)
    predictions["Predicted_Rk"] = list(range(1,predictions.shape[0]+1))
    predictions = predictions.sort_values("Share", ascending=False)
    predictions["Rk"] = list(range(1,predictions.shape[0]+1))
    predictions["Diff"] = (predictions["Rk"] - predictions["Predicted_Rk"])
    return predictions

In [64]:
add_ranks(all_predictions[1])

Player  Share  predictions  Predicted_Rk   Rk  Diff
1530       Karl Malone  0.857     0.192318             2    1    -1
10161   Michael Jordan  0.832     0.167629             3    2    -1
871         Grant Hill  0.327     0.128646             6    3    -3
4487      Tim Hardaway  0.207     0.059984            20    4   -16
7943         Glen Rice  0.117     0.033110            53    5   -48
...                ...    ...          ...           ...  ...   ...
9773    Horacio Llamas  0.000     0.010171           156  437   281
3454     Ennis Whatley  0.000     0.010250           155  438   283
10231  Kevin Salvadori  0.000     0.010553           154  439   285
1084    Aaron Williams  0.000     0.010594           153  440   287
6540     Melvin Booker  0.000    -0.055185           441  441     0

[441 rows x 6 columns]

# Implementing backtesting to predict each year

In [59]:
def backtest(stats, model, years, predictors):
    aps = []
    all_predictions = []
    for year in years:
        train = stats[stats["Year"] < year]
        test = stats[stats["Year"] == year]
        model.fit(train[predictors],train["Share"])
        predictions = model.predict(test[predictors])
        predictions = pd.DataFrame(predictions, columns=["predictions"], index=test.index)
        combination = pd.concat([test[["Player", "Share"]], predictions], axis=1)
        combination = add_ranks(combination)
        all_predictions.append(combination)
        aps.append(find_ap(combination))
    return sum(aps) / len(aps), aps, pd.concat(all_predictions)

In [65]:
mean_ap, aps, all_predictions = backtest(stats, reg, years[5:], predictors)

In [66]:
mean_ap

0.7091945189766616

In [67]:
all_predictions[all_predictions["Rk"] < 5].sort_values("Diff").head(10)

Player  Share  predictions  Predicted_Rk  Rk  Diff
1170        Jason Kidd  0.712     0.028210            52   2   -50
4980        Steve Nash  0.839     0.034100            45   1   -44
8211   Peja Stojaković  0.228     0.036270            38   4   -34
12241      Joakim Noah  0.258     0.046968            37   4   -33
4998        Steve Nash  0.739     0.054129            34   1   -33
5013        Steve Nash  0.785     0.074421            21   2   -19
4487      Tim Hardaway  0.207     0.059984            20   4   -16
900      Allen Iverson  0.270     0.072120            14   4   -10
6494       Kobe Bryant  0.291     0.078014            14   4   -10
1103       Gary Payton  0.372     0.076708            13   3   -10

In [68]:
pd.concat([pd.Series(reg.coef_), pd.Series(predictors)], axis=1).sort_values(0, ascending=False)

0     1
13  0.055322  eFG%
18  0.035209   DRB
29  0.025842  W/L%
17  0.022591   ORB
10  0.016478    2P
22  0.012048   BLK
21  0.011805   STL
15  0.009191   FTA
5   0.007608   FGA
20  0.007590   AST
4   0.007390    FG
7   0.006914    3P
25  0.005624   PTS
12  0.001301   2P%
0   0.000319   Age
30  0.000278    GB
27  0.000111     W
1   0.000091     G
2  -0.000020    GS
26 -0.000166  Year
32 -0.000301  PA/G
28 -0.000334     L
31 -0.000482  PS/G
33 -0.000701   SRS
24 -0.002694    PF
14 -0.003602    FT
3  -0.004189    MP
16 -0.005568   FT%
9  -0.009566   3P%
23 -0.009782   TOV
8  -0.013184   3PA
11 -0.019639   2PA
19 -0.027787   TRB
6  -0.118577   FG%

# Adding more predictors for improving the model

In [69]:
stat_ratios = stats[["PTS", "AST", "STL", "BLK", "3P", "Year"]].groupby("Year").apply(lambda x: x/x.mean())

C:\Users\Pranav Handa\AppData\Local\Temp\ipykernel_19032\2022889540.py:1: FutureWarning: Not prepending group keys to the result index of transform-like apply. In the future, the group keys will be included in the index, regardless of whether the applied function returns a like-indexed object.
To preserve the previous behavior, use

	>>> .groupby(..., group_keys=False)

To adopt the future behavior and silence this warning, use 

	>>> .groupby(..., group_keys=True)
  stat_ratios = stats[["PTS", "AST", "STL", "BLK", "3P", "Year"]].groupby("Year").apply(lambda x: x/x.mean())


In [71]:
stats[["PTS_R", "AST_R", "STL_R", "BLK_R", "3P_R"]] = stat_ratios[["PTS", "AST", "STL", "BLK", "3P"]]

In [72]:
predictors += ["PTS_R", "AST_R", "STL_R", "BLK_R", "3P_R"]

In [73]:
mean_ap, aps, all_predictions = backtest(stats, reg, years[5:], predictors)

# Model Efficiency:

In [74]:
mean_ap

0.7180594999835173

## Ridge regression model gave us an efficiency of 71.8 %